### ISSUE TITLE
Cycle error when renaming/removing a dependency of a resource with create_before_destroy lifecycle

### ISSUE THREAD

**Issue by impl** [hashicorp/terraform#24947]
### Terraform Version

```
Terraform v0.12.24
+ provider.local v1.4.0
+ provider.random v2.2.1
```

### Terraform Configuration Files

Initial:

```hcl
resource "random_id" "id-a" {
  byte_length = 4
}

resource "local_file" "output" {
  content  = random_id.id-a.hex
  filename = "${path.module}/output-${random_id.id-a.hex}"

  lifecycle {
    create_before_destroy = true
  }
}
```

Updated:

```hcl
resource "random_id" "id-b" {
  byte_length = 4
}

resource "local_file" "output" {
  content  = random_id.id-b.hex
  filename = "${path.module}/output-${random_id.id-b.hex}"

  lifecycle {
    create_before_destroy = true
  }
}
```

### Debug Output

[Attached](https://gist.githubusercontent.com/impl/a42188242041e1570fd1378df2582bfb/raw/0da1c204869926d76ec15b3f72093d0996ad0b9e/gistfile1.txt) (note: resource IDs are different here than in the reproduction steps below)

### Expected Behavior

I think it should be the case that removing a dependency of a resource with a `create_before_destroy` lifecycle should move that resource's destroy phase to after the destroy phase of its dependent. However, the graph doesn't appear to capture this.

### Actual Behavior

We encounter an `Error: Cycle` pointing at the destruction of the dependency. Specifically targeting the resource with the `create_before_destroy` lifecycle works around the problem by deferring the analysis of the destruction of the dependency.

### Steps to Reproduce

1. Create initial content:
    ```console
    % cat >main.tf
    resource "random_id" "id-a" {
      byte_length = 4
    }

    resource "local_file" "output" {
      content  = random_id.id-a.hex
      filename = "${path.module}/output-${random_id.id-a.hex}"

      lifecycle {
        create_before_destroy = true
      }
    }
    ^D
    ```
2. Apply:
    ```console
    % terraform init >/dev/null && terraform apply -auto-approve
    random_id.id-a: Creating...
    random_id.id-a: Creation complete after 0s [id=fXyHVw]
    local_file.output: Creating...
    local_file.output: Creation complete after 0s [id=8d15d1a96e106ea74093de4a67d2feed6f4dd077]

    Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
    ```
3. Update content, replacing a dependency:
    ```console
    % sed -i 's/id-a/id-b/g' main.tf
    ```
4. Try to apply again:
    ```console
    % terraform apply -auto-approve
    random_id.id-a: Refreshing state... [id=fXyHVw]
    local_file.output: Refreshing state... [id=8d15d1a96e106ea74093de4a67d2feed6f4dd077]

    Error: Cycle: local_file.output (destroy deposed 732e03e9), random_id.id-a (destroy), local_file.output
    ```

### Additional Context

We observed this issue with the same kind of dependency, a `random_id` being used as a name suffix, but on a `google_container_node_pool` resource. I simplified the problem to this example to avoid having to copy our infrastructure code over here, but I'd be happy to provide it as well if it would make debugging this issue easier. Also, it only seems to make a difference for certain attributes: omitting the `random_id` reference from the `content` in this example does not cause the issue (even though it's still present in the `filename`).

You can work around this bug by targeting the resource with the create-before-destroy lifecycle first. E.g., this works:

```console
% terraform apply -auto-approve -target=local_file.output && terraform apply -auto-approve
local_file.output: Refreshing state... [id=8d15d1a96e106ea74093de4a67d2feed6f4dd077]
random_id.id-b: Creating...
random_id.id-b: Creation complete after 0s [id=JyIw2Q]
local_file.output: Creating...
local_file.output: Creation complete after 0s [id=b2f5d8fa50456e4fbcaf017c85dde7b99cc42fcc]
local_file.output: Destroying... [id=8d15d1a96e106ea74093de4a67d2feed6f4dd077]
local_file.output: Destruction complete after 0s

Warning: Resource targeting is in effect

You are creating a plan with the -target option, which means that the result
of this plan may not represent all of the changes requested by the current
configuration.

The -target option is not for routine use, and is provided only for
exceptional situations such as recovering from errors or mistakes, or when
Terraform specifically suggests to use it as part of an error message.

Warning: Applied changes may be incomplete

The plan was created with the -target option in effect, so some changes
requested in the configuration may have been ignored and the output values may
not be fully updated. Run the following command to verify that no other
changes are pending:
    terraform plan

Note that the -target option is not suitable for routine use, and is provided
only for exceptional situations such as recovering from errors or mistakes, or
when Terraform specifically suggests to use it as part of an error message.

Apply complete! Resources: 2 added, 0 changed, 1 destroyed.
random_id.id-b: Refreshing state... [id=JyIw2Q]
random_id.id-a: Refreshing state... [id=fXyHVw]
local_file.output: Refreshing state... [id=b2f5d8fa50456e4fbcaf017c85dde7b99cc42fcc]
random_id.id-a: Destroying... [id=fXyHVw]
random_id.id-a: Destruction complete after 0s

Apply complete! Resources: 0 added, 0 changed, 1 destroyed.
```

### References

- #21662

**Comment by isaacgraf** [comment]
We are having a very similar issue. We are using aws_sagemaker resources in terraform and we configure create_before_destroy on the aws_sagemaker_endpoint_configuration. If we delete a model inside the endpoint_configuration we get the Cycle Error.
This worked for us in 0.12.6 and broke in 0.12.18 (and still broken in 0.12.25).

Based on the reference above (https://github.com/hashicorp/terraform/issues/21662) I suspect it broke in 0.12.14 with the fix provided for the problem described in that issue.

**Comment by jbardin** [comment]
I've confirmed this is working in master, so the fix will be part of 0.13 release.

**Comment by danieldreier** [comment]
[Terraform 0.13.0 beta 1 launched today](https://github.com/hashicorp/terraform/issues/25016), and you're welcome to confirm the issue you reported is fixed in it. If it isn't, please feel free to file a new issue or reach to to me (ddreier@hashicorp.com) and I can re-open this one.
